pkg/controller/kubebuilderx/controller.go

```diff
@@ -94,22 +98,46 @@ func (c *controller) Do(reconcilers ...Reconciler) Controller {
 }
 
 func (c *controller) Commit() error {
+	defer c.emitFailureEvent()
+
 	if c.err != nil {
 		return c.err
 	}
 	if c.oldTree.GetRoot() == nil {
 		return nil
 	}
 	builder := NewPlanBuilder(c.ctx, c.cli, c.oldTree, c.tree, c.recorder, c.logger)
-	if err := builder.Init(); err != nil {
-		return err
+	if c.err = builder.Init(); c.err != nil {
+		return c.err
+	}
+	var plan graph.Plan
+	plan, c.err = builder.Build()
+	if c.err != nil {
+		return c.err
+	}
+	c.err = plan.Execute()
+	return c.err
+}
+
+func (c *controller) emitFailureEvent() {
+	if c.err == nil {
+		return
+	}
+	if c.tree == nil {
+		return
+	}
+	if c.tree.EventRecorder == nil {
+		return
+	}
+	if c.tree.GetRoot() == nil {
+		return
 	}
-	plan, err := builder.Build()
-	if err != nil {
-		return err
+	// ignore object update optimistic lock conflict
+	if strings.Contains(c.err.Error(), "the object has been modified; please apply your changes to the latest version and try again") {
```

Comment:
replace with `apierrors.IsConflict(err)`

Reply:
fixed in [edb6a85](https://github.com/apecloud/kubeblocks/pull/7336/commits/edb6a857875e680a6eb3197ba66e7dbd8cdb15d7)